src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,730 @@
+                        if let ty::TyFnDef(callee_def_id, substs, _) = f.ty.sty {
+                            callsites.push(CallSite {
+                                caller: def_id,
+                                callee: callee_def_id,
+                                substs: substs,
+                                bb: bb,
+                                location: terminator.source_info
+                            });
+                        }
+                    }
+                }
+
+                in_scc.insert(def_id);
+            }
+        }
+
+        // Move callsites that are in the the SCC to the end so
+        // they're inlined after calls to outside the SCC
+        let mut first_call_in_scc = callsites.len();
+
+        let mut i = 0;
+        while i < first_call_in_scc {
+            let f = callsites[i].caller;
+            if in_scc.contains(&f) {
+                first_call_in_scc -= 1;
+                callsites.swap(i, first_call_in_scc);
+            } else {
+                i += 1;
+            }
+        }
+
+        let mut local_change;
+        let mut changed = false;
+
+        loop {
+            local_change = false;
+            let mut csi = 0;
+            while csi < callsites.len() {
+                let foreign_mir;
+
+                let callsite = callsites[csi];
+                csi += 1;
+
+                let callee_mir = {
+                    let callee_mir : Option<&Mir<'tcx>> = if callsite.callee.is_local() {
+                        map.map.get(&callsite.callee)
+                    } else {
+                        foreign_mir = self.get_foreign_mir(callsite.callee);
+                        foreign_mir.as_ref().map(|m| &**m)
+                    };
+
+                    let callee_mir = if let Some(m) = callee_mir {
+                        m
+                    } else {
+                        continue;
+                    };
+
+                    if !self.should_inline(callsite, callee_mir) {
+                        continue;
+                    }
+
+                    callee_mir.subst(self.tcx, callsite.substs)
+                };
+
+                let caller_mir = map.map.get_mut(&callsite.caller).unwrap();
+
+                let start = caller_mir.basic_blocks().len();
+
+                if !self.inline_call(callsite, caller_mir, callee_mir) {
+                    continue;
+                }
+
+                // Add callsites from inlined function
+                for (bb, bb_data) in caller_mir.basic_blocks().iter_enumerated().skip(start) {
+                    // Only consider direct calls to functions
+                    let terminator = bb_data.terminator();
+                    if let TerminatorKind::Call {
+                        func: Operand::Constant(ref f), .. } = terminator.kind {
+                        if let ty::TyFnDef(callee_def_id, substs, _) = f.ty.sty {
+                            // Don't inline the same function multiple times.
```

Comment:
I'm not sure it's even cost-effective across the board to inline non-leaves without a more complex analysis.